cmd/load-generator/boulder-calls.go

```diff
@@ -0,0 +1,482 @@
+			terms = l[1 : len(l)-len(">;rel=\"terms-of-service\"")]
+			break
+		}
+	}
+
+	// agree to terms
+	regStr = []byte(fmt.Sprintf(`{"resource":"reg","agreement":"%s"}`, terms))
+
+	// build the JWS object
+	requestPayload, err = s.signWithNonce("/acme/reg", false, regStr, signer)
+	if err != nil {
+		return fmt.Errorf("/acme/reg, sign failed: %s", err)
+	}
+
+	tStarted := time.Now()
+	resp, err = s.post(resp.Header.Get("Location"), requestPayload, ctx.ns)
+	tFinished := time.Now()
+	tState := "good"
+	defer func() { s.callLatency.Add("POST /acme/reg/{ID}", tStarted, tFinished, tState) }()
+	if err != nil {
+		tState = "error"
+		return fmt.Errorf("/acme/reg, post failed: %s", err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 202 {
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			// just fail
+			tState = "error"
+			return err
+		}
+		tState = "error"
+		return fmt.Errorf("/acme/reg, bad response status %d: %s", resp.StatusCode, body)
+	}
+
+	ctx.reg = &registration{key: signKey, signer: signer}
+	s.addRegistration(ctx.reg)
+	return nil
+}
+
+func newAuthorization(s *State, ctx *context) error {
+	// generate a random(-ish) domain name,  will cause some multiples but not enough to make rate limits annoying!
+	n := time.Now().UnixNano()
+	b := new(bytes.Buffer)
+	binary.Write(b, binary.LittleEndian, n)
+	randomDomain := fmt.Sprintf("%x.%s", sha1.Sum(b.Bytes()), s.domainBase)
+
+	// create the registration object
+	initAuth := fmt.Sprintf(`{"resource":"new-authz","identifier":{"type":"dns","value":"%s"}}`, randomDomain)
+
+	// build the JWS object
+	getNew := false
+	if mrand.Intn(1) == 0 {
+		getNew = true
+	}
+	requestPayload, err := s.signWithNonce("/acme/new-authz", getNew, []byte(initAuth), ctx.reg.signer)
+	if err != nil {
+		return err
+	}
+
+	started := time.Now()
+	resp, err := s.post(fmt.Sprintf("%s/acme/new-authz", s.apiBase), requestPayload, ctx.ns)
+	finished := time.Now()
+	state := "good"
+	defer func() { s.callLatency.Add("POST /acme/new-authz", started, finished, state) }()
+	if err != nil {
+		state = "error"
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 201 {
+		// something
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			// just fail
+			state = "error"
+			return err
+		}
+		state = "error"
+		return fmt.Errorf("bad response, status %d: %s", resp.StatusCode, body)
+	}
+	// location := resp.Header.Get("Location")
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		state = "error"
+		// just fail
+		return err
+	}
+
+	var authz core.Authorization
+	err = json.Unmarshal(body, &authz)
+	if err != nil {
+		state = "error"
+		return err
+	}
+	// populate authz ID from location header because we strip it
+	paths := strings.Split(resp.Header.Get("Location"), "/")
+	authz.ID = paths[len(paths)-1]
+
+	ctx.pendingAuthz = append(ctx.pendingAuthz, &authz)
+	return nil
+}
+
+func getPending(ctx *context) *core.Authorization {
```

Comment:
Maybe this should be called `popPending` or something that would make it clearer that its a "destructive" operation in that it seems to remove the selected pending authz from the `ctx`'s `pendingAuthz` slice.